dbms/src/Interpreters/Expand.cpp

```diff
@@ -0,0 +1,240 @@
+#include <tipb/executor.pb.h>
+
+#include "DataTypes/DataTypesNumber.h"
+
+namespace DB
+{
+
+namespace /// anonymous namespace for storing private function utils.
+{
+void convertColumnToNullable(ColumnWithTypeAndName & column)
+{
+    column.type = makeNullable(column.type);
+    if (column.column)
+        column.column = makeNullable(column.column);
+}
+} // namespace
+
+Expand::Expand(const DB::GroupingSets & gss)
+    : group_sets_names(gss)
+{}
+
+void Expand::getGroupingSetsDes(FmtBuffer & buffer) const
+{
+    buffer.append("[");
+    for (const auto & grouping_set : group_sets_names)
+    {
+        buffer.append("{");
+        for (const auto & grouping_exprs : grouping_set)
+        {
+            buffer.append("<");
+            buffer.joinStr(grouping_exprs.begin(), grouping_exprs.end());
+            buffer.append(">");
+        }
+        buffer.append("}");
+    }
+    buffer.append("]");
+}
+
+/// for cases like: select count(distinct a), count(distinct b) from t;
+/// it will generate 2 group set with <a> and <b>, over which we should
+/// expand one more replica of the source rows from the input block and
+/// identify it with the grouping id in the appended new column.
+///
+/// eg: source block         ==>        replicated block
+///      <a, b>              ==>         <a, b, groupingID>  a new column is appended
+///      1  1       target a -+----->     1  null  groupingID for a =1
+///      2  2                 +----->     2  null  groupingID for b =2
+///                 target b -+----->     null  1  groupingID for a =1
+///                           +----->     null  2  groupingID for b =2
+///
+/// when target a specified group set, other group set columns should be filled
+/// with null value to make group by(a,b) operator to meet the equivalence effect
+/// of group by(a) and group by(b) since the other group set columns has been filled
+/// with null value.
+///
+/// \param input the source block
+/// \return
+
+void Expand::replicateAndFillNull(Block & block) const
+{
+    size_t origin_rows = block.rows();
+    // make a replicate slice, using it to replicate origin rows.
+    std::unique_ptr<IColumn::Offsets> offsets_to_replicate;
+    offsets_to_replicate = std::make_unique<IColumn::Offsets>(origin_rows);
+
+    // get the replicate offset fixed as group set num.
+    IColumn::Offset current_offset = 0;
+    const IColumn::Offset replicate_times_for_one_row = getGroupSetNum();
+
+    // create a column for grouping id.
+    auto grouping_id_column = ColumnUInt64::create();
+    auto & grouping_id_column_data = grouping_id_column->getData();
+    // reserve N times of current block rows size.
+    grouping_id_column_data.reserve(block.rows() * replicate_times_for_one_row);
+
+    // prepare added mutable grouping id column.
+    MutableColumns added_grouping_id_column;
+    added_grouping_id_column.reserve(1);
+    added_grouping_id_column.push_back(grouping_id_column->getPtr());
```

Comment:
Seems useless here? why not use `grouping_id_column` directly?

Reply:
seem works, I ref-ed some other similar code before